File: portfolio/projects/D_INSA/INSA/frontend/src/api/travelOrder.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import apiClient from "./client";

export type TravelType = "DOMESTIC" | "OVERSEAS";

export interface CompanionInfo {
  emp_id: number;
  emp_no: string | null;
  name_ko: string | null;
  dept_name: string | null;
}

export interface TravelOrderCreateBody {
  line_template_id: number;
  title?: string | null;
  travel_type: TravelType;
  purpose: string;
  destination: string;
  client_company?: string | null;
  start_at: string;
  end_at: string;
  transportation?: string | null;
  estimated_cost?: string | null;
  project_code?: string | null;
  appraisal_case_no?: string | null;
  remarks?: string | null;
  companion_emp_ids: number[];
}

export interface TravelOrderDetailResponse {
  id: number;
  doc_id: number;
  travel_type: string;
  purpose: string;
  destination: string;
  client_company: string | null;
  start_at: string;
  end_at: string;
  transportation: string | null;
  estimated_cost: string | null;
  project_code: string | null;
  appraisal_case_no: string | null;
  remarks: string | null;
  created_at: string | null;
  companions: CompanionInfo[];
}

export interface TravelOrderListRow {
  doc_id: number;
  doc_no: string | null;
  title: string;
  status: string;
  drafter_id: number;
  drafter_name: string | null;
  drafter_emp_no: string | null;
  dept_name: string | null;
  travel_type: string;
  destination: string;
  purpose: string;
  start_at: string;
  end_at: string;
  appraisal_case_no: string | null;
  estimated_cost: string | null;
  companion_count: number;
  drafted_at: string | null;
  completed_at: string | null;
  has_report: boolean;
}

function invalidateTravel(qc: ReturnType<typeof useQueryClient>) {
  qc.invalidateQueries({ queryKey: ["travel"] });
  qc.invalidateQueries({ queryKey: ["approval"] });
  qc.invalidateQueries({ queryKey: ["calendar"] });
}

// ── Mutations ──────────────────────────────────────────────

export function useCreateTravelOrder() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: TravelOrderCreateBody) => {
      const { data } = await apiClient.post<{
        doc_id: number;
        detail_id: number;
        status: string;
      }>("/travel/orders", body);
      return data;
    },
    onSuccess: () => invalidateTravel(qc),
  });
}

export function useSubmitTravelOrder() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (docId: number) => {
      const { data } = await apiClient.post<{ doc_id: number; status: string }>(
        `/travel/orders/${docId}/submit`,
      );
      return data;
    },
    onSuccess: () => invalidateTravel(qc),
  });
}

export function useCancelTravelOrder() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (docId: number) => {
      const { data } = await apiClient.post<{ doc_id: number; status: string }>(
        `/travel/orders/${docId}/cancel`,
      );
      return data;
    },
    onSuccess: () => invalidateTravel(qc),
  });
}

// ── Queries ────────────────────────────────────────────────

export function useMyTravelOrders(params?: {
  status?: string;
  include_companion?: boolean;
}) {
  return useQuery({
    queryKey: [
      "travel",
      "orders",
      "my",
      params?.status ?? "all",
      params?.include_companion ?? true,
    ],
    queryFn: async () => {
      const { data } = await apiClient.get<TravelOrderListRow[]>(
        "/travel/orders/my",
        {
          params: {
            ...(params?.status ? { status: params.status } : {}),
            include_companion: params?.include_companion ?? true,
          },
        },
      );
      return data;
    },
  });
}

export function useTeamTravelOrders(params?: { status?: string }) {
  return useQuery({
    queryKey: ["travel", "orders", "team", params?.status ?? "all"],
    queryFn: async () => {
      const { data } = await apiClient.get<TravelOrderListRow[]>(
        "/travel/orders/team",
        {
          params: params?.status ? { status: params.status } : {},
        },
      );
      return data;
    },
  });
}

export function useTravelOrderDetail(docId: number | null) {
  return useQuery({
    queryKey: ["travel", "orders", "detail", docId],
    queryFn: async () => {
      const { data } = await apiClient.get<TravelOrderDetailResponse>(
        `/travel/orders/${docId}`,
      );
      return data;
    },
    enabled: docId !== null,
  });
}

export function useTravelOrdersByCaseNo(caseNo: string | null) {
  return useQuery({
    queryKey: ["travel", "orders", "by-case", caseNo],
    queryFn: async () => {
      const { data } = await apiClient.get<TravelOrderListRow[]>(
        `/travel/orders/by-case/${encodeURIComponent(caseNo ?? "")}`,
      );
      return data;
    },
    enabled: !!caseNo,
  });
}

File: portfolio/projects/D_INSA/INSA/backend/app/schemas/travel_order.py
"""Travel order / report schemas (PHASE 16)."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

TravelType = Literal["DOMESTIC", "OVERSEAS"]


class TravelOrderCreate(BaseModel):
    line_template_id: int
    title: Optional[str] = Field(default=None, max_length=200)
    travel_type: TravelType = "DOMESTIC"
    purpose: str = Field(..., max_length=500)
    destination: str = Field(..., max_length=200)
    client_company: Optional[str] = Field(default=None, max_length=200)
    start_at: datetime
    end_at: datetime
    transportation: Optional[str] = Field(default=None, max_length=100)
    estimated_cost: Optional[Decimal] = None
    project_code: Optional[str] = Field(default=None, max_length=50)
    appraisal_case_no: Optional[str] = Field(default=None, max_length=100)
    remarks: Optional[str] = Field(default=None, max_length=500)
    companion_emp_ids: list[int] = Field(default_factory=list)


class CompanionInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    emp_id: int
    emp_no: Optional[str] = None
    name_ko: Optional[str] = None
    dept_name: Optional[str] = None


class TravelOrderDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    doc_id: int
    travel_type: str
    purpose: str
    destination: str
    client_company: Optional[str] = None
    start_at: datetime
    end_at: datetime
    transportation: Optional[str] = None
    estimated_cost: Optional[Decimal] = None
    project_code: Optional[str] = None
    appraisal_case_no: Optional[str] = None
    remarks: Optional[str] = None
    created_at: Optional[datetime] = None
    companions: list[CompanionInfo] = Field(default_factory=list)


class TravelOrderListRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    doc_id: int
    doc_no: Optional[str] = None
    title: str
    status: str
    drafter_id: int
    drafter_name: Optional[str] = None
    drafter_emp_no: Optional[str] = None
    dept_name: Optional[str] = None
    travel_type: str
    destination: str
    purpose: str
    start_at: datetime
    end_at: datetime
    appraisal_case_no: Optional[str] = None
    estimated_cost: Optional[Decimal] = None
    companion_count: int = 0
    drafted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    has_report: bool = False


# ── Travel report (복명서) ────────────────────────────────


class TravelReportCreate(BaseModel):
    line_template_id: int
    title: Optional[str] = Field(default=None, max_length=200)
    report_content: str = Field(..., min_length=1)
    actual_cost: Optional[Decimal] = None
    receipts_url: Optional[str] = Field(default=None, max_length=500)


class TravelReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    travel_id: int
    doc_id: Optional[int] = None
    report_content: str
    actual_cost: Optional[Decimal] = None
    receipts_url: Optional[str] = None
    reported_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
